packages/core/src/components/fields/ArrayField.tsx

```diff
@@ -72,210 +68,827 @@ function keyedToPlainFormData<T>(keyedFormData: KeyedFormDataType<T> | KeyedForm
-    return schemaUtils.getDefaultFormState(itemSchema) as unknown as T;
+      const itemProps = {
+        index,
+        itemKey: key,
+        name: name && `${name}-${index}`,
+        registry,
+        uiOptions,
+        hideError,
+        readonly,
+        disabled,
+        required,
+        title: fieldTitle ? `${fieldTitle}-${index + 1}` : undefined,
+        canAdd,
+        canRemove: additional,
+        canMoveUp: index >= itemSchemas.length + 1,
+        canMoveDown: additional && index < items.length - 1,
+        itemSchema,
+        itemData: itemCast,
+        itemUiSchema,
+        itemFieldPathId,
+        itemErrorSchema,
+        autofocus: autofocus && index === 0,
+        onBlur,
+        onFocus,
+        rawErrors,
+        totalItems: keyedFormData.length,
+        onChange,
+        handleAddItem,
+        handleCopyItem,
+        handleRemoveItem,
+        handleReorderItems,
+      };
+      return <ArrayFieldItem key={key} {...itemProps} />;
+    }),
+    onAddClick: handleAddItem,
+    readonly,
+    required,
+    registry,
+    schema,
+    uiSchema,
+    title: fieldTitle,
+    errorSchema,
+    rawErrors,
+    optionalDataControl,
   };
 
+  const Template = getTemplate<'ArrayFieldTemplate', T[], S, F>('ArrayFieldTemplate', registry, uiOptions);
+  return <Template {...arrayProps} />;
+}
+
+interface KeyedFormDataState<T = any> {
+  /** The keyed form data elements */
+  keyedFormData: KeyedFormDataType<T>[];
+  /** Updates the keyed form data elements to the given value */
+  updateKeyedFormData: (newData: KeyedFormDataType<T>[]) => T[];
+}
+
+/** Type used for the state of the `ArrayField` component */
+type ArrayFieldState<T> = {
+  /** The hash of the last formData passed in */
+  formDataHash: string;
+  /** The keyed form data elements */
+  keyedFormData: KeyedFormDataType<T>[];
+};
+
+/** A custom hook that handles the updating of the keyedFormData from an external `formData` change as well as
+ * internally by the `ArrayField`. If there was an external `formData` change, then the `keyedFormData` is recomputed
+ * in order to preserve the unique keys from the old `keyedFormData` to the new `formData`. Along with the
+ * `keyedFormData` this hook also returns an `updateKeyedFormData()` function for use by the `ArrayField`. The detection
+ * of external `formData` are handled by storing the hash of that `formData` along with the `keyedFormData` associated
+ * with it. The `updateKeyedFormData()` will update that hash whenever the `keyedFormData` is modified and as well as
+ * returning the plain `formData` from the `keyedFormData`.
+ */
+function useKeyedFormData<T = any>(formData: T[] = []): KeyedFormDataState<T> {
+  const [state, setState] = useState<ArrayFieldState<T>>({
+    formDataHash: hashObject(formData),
+    keyedFormData: generateKeyedFormData<T>(formData),
+  });
+
+  let { keyedFormData, formDataHash } = state;
+  const newHash = hashObject(formData);
```

Comment:
Let's at least memoize the hash so we don't need to recompute it every render. We can also 'lazy-initialize` state so `generateKeyedFormData` only runs once.
```suggestion
  const newHash = useMemo(() => hashObject(formData), [formData])
  const [state, setState] = useState<ArrayFieldState<T>>(() => ({
    formDataHash: newHash,
    keyedFormData: generateKeyedFormData<T>(formData),
  }));

  let { keyedFormData, formDataHash } = state;
```